File: mobile/modules/bluetooth-sdk/ios/Source/streaming/StreamModels.swift
import Foundation

public struct StreamVideoConfig {
    public let width: Int?
    public let height: Int?
    public let bitrate: Int?
    public let fps: Int?

    public init(
        width: Int? = nil,
        height: Int? = nil,
        bitrate: Int? = nil,
        fps: Int? = nil
    ) {
        self.width = width
        self.height = height
        self.bitrate = bitrate
        self.fps = fps
    }

    var dictionary: [String: Any] {
        var values: [String: Any] = [:]
        if let width { values["width"] = width }
        if let height { values["height"] = height }
        if let bitrate { values["bitrate"] = bitrate }
        // ASG stream parsers shipped with the BLE key named "frameRate".
        if let fps { values["frameRate"] = fps }
        return values
    }

    init?(values: [String: Any]?) {
        guard let values else { return nil }
        self.init(
            width: intValue(values["width"]),
            height: intValue(values["height"]),
            bitrate: intValue(values["bitrate"]),
            fps: intValue(values["fps"])
        )
    }
}

public struct StreamAudioConfig {
    public let bitrate: Int?
    public let sampleRate: Int?
    public let echoCancellation: Bool?
    public let noiseSuppression: Bool?

    public init(
        bitrate: Int? = nil,
        sampleRate: Int? = nil,
        echoCancellation: Bool? = nil,
        noiseSuppression: Bool? = nil
    ) {
        self.bitrate = bitrate
        self.sampleRate = sampleRate
        self.echoCancellation = echoCancellation
        self.noiseSuppression = noiseSuppression
    }

    var dictionary: [String: Any] {
        var values: [String: Any] = [:]
        if let bitrate { values["bitrate"] = bitrate }
        if let sampleRate { values["sampleRate"] = sampleRate }
        if let echoCancellation { values["echoCancellation"] = echoCancellation }
        if let noiseSuppression { values["noiseSuppression"] = noiseSuppression }
        return values
    }

    init?(values: [String: Any]?) {
        guard let values else { return nil }
        self.init(
            bitrate: intValue(values["bitrate"]),
            sampleRate: intValue(values["sampleRate"]),
            echoCancellation: values["echoCancellation"] as? Bool,
            noiseSuppression: values["noiseSuppression"] as? Bool
        )
    }
}

/// Effective video settings reported by the glasses after defaults and clamps.
public struct StreamResolvedVideoConfig: Equatable {
    /// Encoded output width sent to the stream endpoint.
    public let width: Int
    /// Encoded output height sent to the stream endpoint.
    public let height: Int
    /// Native camera buffer width selected before crop/downscale.
    public let captureWidth: Int?
    /// Native camera buffer height selected before crop/downscale.
    public let captureHeight: Int?
    /// Encoded video bitrate in bits per second.
    public let bitrate: Int
    /// Resolved capture/encode frame rate.
    public let fps: Double

    public init(
        width: Int,
        height: Int,
        captureWidth: Int? = nil,
        captureHeight: Int? = nil,
        bitrate: Int,
        fps: Double
    ) {
        self.width = width
        self.height = height
        self.captureWidth = captureWidth
        self.captureHeight = captureHeight
        self.bitrate = bitrate
        self.fps = fps
    }

    init?(values: [String: Any]?) {
        guard let values,
              let width = intValue(values["width"]),
              let height = intValue(values["height"]),
              let bitrate = intValue(values["bitrate"]),
              let fps = doubleValue(values["fps"])
        else {
            return nil
        }
        self.init(
            width: width,
            height: height,
            captureWidth: intValue(values["captureWidth"]),
            captureHeight: intValue(values["captureHeight"]),
            bitrate: bitrate,
            fps: fps
        )
    }

    var values: [String: Any] {
        var values: [String: Any] = [
            "width": width,
            "height": height,
            "bitrate": bitrate,
            "fps": fps,
        ]
        if let captureWidth { values["captureWidth"] = captureWidth }
        if let captureHeight { values["captureHeight"] = captureHeight }
        return values
    }
}

public struct StreamResolvedAudioConfig: Equatable {
    public let bitrate: Int?
    public let sampleRate: Int?
    public let echoCancellation: Bool?
    public let noiseSuppression: Bool?

    public init(
        bitrate: Int? = nil,
        sampleRate: Int? = nil,
        echoCancellation: Bool? = nil,
        noiseSuppression: Bool? = nil
    ) {
        self.bitrate = bitrate
        self.sampleRate = sampleRate
        self.echoCancellation = echoCancellation
        self.noiseSuppression = noiseSuppression
    }

    init?(values: [String: Any]?) {
        guard let values else { return nil }
        self.init(
            bitrate: intValue(values["bitrate"]),
            sampleRate: intValue(values["sampleRate"]),
            echoCancellation: boolValue(values, "echoCancellation"),
            noiseSuppression: boolValue(values, "noiseSuppression")
        )
    }

    var values: [String: Any] {
        var values: [String: Any] = [:]
        if let bitrate { values["bitrate"] = bitrate }
        if let sampleRate { values["sampleRate"] = sampleRate }
        if let echoCancellation { values["echoCancellation"] = echoCancellation }
        if let noiseSuppression { values["noiseSuppression"] = noiseSuppression }
        return values
    }
}

public enum StreamTransport: String, Equatable {
    case rtmp
    case srt
    case whip
}

public struct StreamResolvedConfig: Equatable {
    public let transport: StreamTransport?
    public let video: StreamResolvedVideoConfig?
    public let audio: StreamResolvedAudioConfig?

    public init(
        transport: StreamTransport? = nil,
        video: StreamResolvedVideoConfig? = nil,
        audio: StreamResolvedAudioConfig? = nil
    ) {
        self.transport = transport
        self.video = video
        self.audio = audio
    }

    init?(values: [String: Any]?) {
        guard let values else { return nil }
        self.init(
            transport: stringValue(values, "transport").flatMap(StreamTransport.init(rawValue:)),
            video: StreamResolvedVideoConfig(values: values["video"] as? [String: Any]),
            audio: StreamResolvedAudioConfig(values: values["audio"] as? [String: Any])
        )
    }

    var values: [String: Any] {
        var values: [String: Any] = [:]
        if let transport { values["transport"] = transport.rawValue }
        if let video { values["video"] = video.values }
        if let audio { values["audio"] = audio.values }
        return values
    }
}

public struct StreamRequest {
    public let streamUrl: String
    public let streamId: String
    public let sound: Bool
    public let video: StreamVideoConfig?
    public let audio: StreamAudioConfig?

    public init(
        streamUrl: String,
        streamId: String = "",
        sound: Bool = true,
        video: StreamVideoConfig? = nil,
        audio: StreamAudioConfig? = nil
    ) {
        self.streamUrl = streamUrl
        self.streamId = streamId
        self.sound = sound
        self.video = video
        self.audio = audio
    }

    init(values: [String: Any]) {
        self.init(
            streamUrl: values["streamUrl"] as? String
                ?? values["rtmpUrl"] as? String
                ?? values["srtUrl"] as? String
                ?? values["whipUrl"] as? String
                ?? "",
            streamId: values["streamId"] as? String ?? "",
            sound: values["sound"] as? Bool ?? true,
            video: StreamVideoConfig(values: values["video"] as? [String: Any]),
            audio: StreamAudioConfig(values: values["audio"] as? [String: Any])
        )
    }

    public var values: [String: Any] {
        var values: [String: Any] = [:]
        values["type"] = "start_stream"
        values["streamUrl"] = streamUrl
        values["streamId"] = streamId
        values["sound"] = sound
        if let videoValues = video?.dictionary, !videoValues.isEmpty {
            values["video"] = videoValues
        }
        if let audioValues = audio?.dictionary, !audioValues.isEmpty {
            values["audio"] = audioValues
        }
        return values
    }
}

struct StreamKeepAliveRequest {
    let streamId: String
    let ackId: String

    init(streamId: String, ackId: String) {
        self.streamId = streamId
        self.ackId = ackId
    }

    init(values: [String: Any]) {
        self.init(
            streamId: values["streamId"] as? String ?? "",
            ackId: values["ackId"] as? String ?? ""
        )
    }

    var values: [String: Any] {
        var values: [String: Any] = [:]
        values["type"] = "keep_stream_alive"
        values["streamId"] = streamId
        values["ackId"] = ackId
        return values
    }
}

public enum StreamState: String, Equatable {
    case initializing
    case streaming
    case stopping
    case stopped
    case reconnecting
    case reconnected
    case reconnectFailed = "reconnect_failed"
    case error

    fileprivate static func from(_ value: String?) -> StreamState? {
        switch value?.lowercased() {
        case "initializing", "starting", "connecting":
            return .initializing
        case "streaming", "streaming_started", "active":
            return .streaming
        case "stopping":
            return .stopping
        case "stopped", "not_streaming", "disconnected", "timeout":
            return .stopped
        case "reconnecting":
            return .reconnecting
        case "reconnected":
            return .reconnected
        case "reconnect_failed":
            return .reconnectFailed
        case "error", "error_not_streaming":
            return .error
        default:
            return nil
        }
    }
}

public enum StreamStatusKind: String, Equatable {
    case lifecycle
    case reconnect
    case error
    case snapshot
}

public enum StreamStatus: CustomStringConvertible, Equatable {
    case lifecycle(state: StreamState, streamId: String?, timestamp: Int?, resolvedConfig: StreamResolvedConfig?)
    case reconnecting(
        streamId: String?,
        attempt: Int,
        maxAttempts: Int,
        reason: String,
        timestamp: Int?,
        resolvedConfig: StreamResolvedConfig?
    )
    case reconnected(streamId: String?, attempt: Int, timestamp: Int?, resolvedConfig: StreamResolvedConfig?)
    case reconnectFailed(streamId: String?, maxAttempts: Int, timestamp: Int?, resolvedConfig: StreamResolvedConfig?)
    case error(streamId: String?, errorDetails: String, timestamp: Int?, resolvedConfig: StreamResolvedConfig?)
    case snapshot(
        state: StreamState,
        streaming: Bool,
        reconnecting: Bool,
        streamId: String?,
        attempt: Int?,
        timestamp: Int?,
        resolvedConfig: StreamResolvedConfig?
    )

    public init(values: [String: Any]) {
        let rawState = stringValue(values, "status")
        let streamId = stringValue(values, "streamId")
        let timestamp = intValue(values["timestamp"])
        let resolvedConfig = StreamResolvedConfig(values: values["resolvedConfig"] as? [String: Any])
        let attempt = optionalIntValue(values, "attempt")
        let maxAttempts = optionalIntValue(values, "maxAttempts") ?? 0
        let parsedState = StreamState.from(rawState)

        if hasAnyKey(values, "streaming") || hasAnyKey(values, "reconnecting") {
            let streaming = boolValue(values, "streaming") == true
            let reconnecting = boolValue(values, "reconnecting") == true
            let snapshotState: StreamState
            if reconnecting {
                snapshotState = .reconnecting
            } else if streaming {
                snapshotState = .streaming
            } else {
                snapshotState = parsedState ?? .stopped
            }
            self = .snapshot(
                state: snapshotState,
                streaming: streaming,
                reconnecting: reconnecting,
                streamId: streamId,
                attempt: attempt,
                timestamp: timestamp,
                resolvedConfig: resolvedConfig
            )
            return
        }

        guard let state = parsedState else {
            self = .error(
                streamId: streamId,
                errorDetails: rawState.map { "Unknown stream status: \($0)" } ?? "Missing stream status",
                timestamp: timestamp,
                resolvedConfig: resolvedConfig
            )
            return
        }

        switch state {
        case .reconnecting:
            self = .reconnecting(
                streamId: streamId,
                attempt: attempt ?? 0,
                maxAttempts: maxAttempts,
                reason: stringValue(values, "reason") ?? "",
                timestamp: timestamp,
                resolvedConfig: resolvedConfig
            )
        case .reconnected:
            self = .reconnected(
                streamId: streamId,
                attempt: attempt ?? 0,
                timestamp: timestamp,
                resolvedConfig: resolvedConfig
            )
        case .reconnectFailed:
            self = .reconnectFailed(
                streamId: streamId,
                maxAttempts: maxAttempts,
                timestamp: timestamp,
                resolvedConfig: resolvedConfig
            )
        case .error:
            self = .error(
                streamId: streamId,
                errorDetails: stringValue(values, "errorDetails")
                    ?? (rawState == "error_not_streaming" ? "not_streaming" : "Unknown stream error"),
                timestamp: timestamp,
                resolvedConfig: resolvedConfig
            )
        default:
            self = .lifecycle(
                state: state,
                streamId: streamId,
                timestamp: timestamp,
                resolvedConfig: resolvedConfig
            )
        }
    }

    public var kind: StreamStatusKind {
        switch self {
        case .lifecycle:
            .lifecycle
        case .reconnecting, .reconnected, .reconnectFailed:
            .reconnect
        case .error:
            .error
        case .snapshot:
            .snapshot
        }
    }

    public var state: StreamState {
        switch self {
        case let .lifecycle(state, _, _, _):
            state
        case .reconnecting:
            .reconnecting
        case .reconnected:
            .reconnected
        case .reconnectFailed:
            .reconnectFailed
        case .error:
            .error
        case let .snapshot(state, _, _, _, _, _, _):
            state
        }
    }

    public var streamId: String? {
        switch self {
        case let .lifecycle(_, streamId, _, _),
             let .reconnecting(streamId, _, _, _, _, _),
             let .reconnected(streamId, _, _, _),
             let .reconnectFailed(streamId, _, _, _),
             let .error(streamId, _, _, _),
             let .snapshot(_, _, _, streamId, _, _, _):
            streamId
        }
    }

    public var timestamp: Int? {
        switch self {
        case let .lifecycle(_, _, timestamp, _),
             let .reconnecting(_, _, _, _, timestamp, _),
             let .reconnected(_, _, timestamp, _),
             let .reconnectFailed(_, _, timestamp, _),
             let .error(_, _, timestamp, _),
             let .snapshot(_, _, _, _, _, timestamp, _):
            timestamp
        }
    }

    public var resolvedConfig: StreamResolvedConfig? {
        switch self {
        case let .lifecycle(_, _, _, resolvedConfig),
             let .reconnecting(_, _, _, _, _, resolvedConfig),
             let .reconnected(_, _, _, resolvedConfig),
             let .reconnectFailed(_, _, _, resolvedConfig),
             let .error(_, _, _, resolvedConfig),
             let .snapshot(_, _, _, _, _, _, resolvedConfig):
            resolvedConfig
        }
    }

    public var values: [String: Any] {
        var values: [String: Any] = [
            "kind": kind.rawValue,
            "status": state.rawValue,
        ]
        if let streamId, !streamId.isEmpty {
            values["streamId"] = streamId
        }
        if let timestamp {
            values["timestamp"] = timestamp
        }
        if let resolvedConfig {
            values["resolvedConfig"] = resolvedConfig.values
        }

        switch self {
        case .lifecycle:
            break
        case let .reconnecting(_, attempt, maxAttempts, reason, _, _):
            values["attempt"] = attempt
            values["maxAttempts"] = maxAttempts
            values["reason"] = reason
        case let .reconnected(_, attempt, _, _):
            values["attempt"] = attempt
        case let .reconnectFailed(_, maxAttempts, _, _):
            values["maxAttempts"] = maxAttempts
        case let .error(_, errorDetails, _, _):
            values["errorDetails"] = errorDetails
        case let .snapshot(_, streaming, reconnecting, _, attempt, _, _):
            values["streaming"] = streaming
            values["reconnecting"] = reconnecting
            if let attempt {
                values["attempt"] = attempt
            }
        }

        return values
    }

    public var description: String {
        "StreamStatus(kind: \(kind.rawValue), status: \(state.rawValue), streamId: \(streamId ?? "none"))"
    }
}

public struct StreamStatusEvent: CustomStringConvertible {
    public let status: StreamStatus

    public init(status: StreamStatus) {
        self.status = status
    }

    public init(values: [String: Any]) {
        status = StreamStatus(values: values)
    }

    public var state: StreamState {
        status.state
    }

    public var streamId: String? {
        status.streamId
    }

    public var resolvedConfig: StreamResolvedConfig? {
        status.resolvedConfig
    }

    public var values: [String: Any] {
        var values = status.values
        values["type"] = "stream_status"
        return values
    }

    public var description: String {
        "StreamStatusEvent(kind: \(status.kind.rawValue), status: \(state.rawValue), streamId: \(streamId ?? "none"))"
    }
}

public struct KeepAliveAckEvent: CustomStringConvertible, Equatable {
    public let streamId: String
    public let ackId: String
    public let timestamp: Int?

    public init(streamId: String, ackId: String, timestamp: Int? = nil) {
        self.streamId = streamId
        self.ackId = ackId
        self.timestamp = timestamp
    }

    public init(values: [String: Any]) {
        streamId = stringValue(values, "streamId") ?? ""
        ackId = stringValue(values, "ackId") ?? ""
        timestamp = intValue(values["timestamp"])
    }

    public var values: [String: Any] {
        var values: [String: Any] = [
            "type": "keep_alive_ack",
            "streamId": streamId,
            "ackId": ackId,
        ]
        if let timestamp {
            values["timestamp"] = timestamp
        }
        return values
    }

    public var description: String {
        "KeepAliveAckEvent(streamId: \(streamId), ackId: \(ackId))"
    }
}

File: mobile/modules/bluetooth-sdk/android/src/main/java/com/mentra/bluetoothsdk/streaming/StreamModels.kt
package com.mentra.bluetoothsdk

data class StreamVideoConfig @JvmOverloads constructor(
    val width: Int? = null,
    val height: Int? = null,
    val bitrate: Int? = null,
    val fps: Int? = null,
) {
    fun toMap(): Map<String, Any> =
        listOfNotNull(
            width?.let { "width" to it },
            height?.let { "height" to it },
            bitrate?.let { "bitrate" to it },
            // ASG stream parsers shipped with the BLE key named "frameRate".
            fps?.let { "frameRate" to it },
        ).toMap()

    companion object {
        @JvmStatic
        fun fromMap(values: Map<String, Any>?): StreamVideoConfig? {
            values ?: return null
            return StreamVideoConfig(
                width = numberValue(values, "width"),
                height = numberValue(values, "height"),
                bitrate = numberValue(values, "bitrate"),
                fps = numberValue(values, "fps"),
            )
        }
    }
}

data class StreamAudioConfig @JvmOverloads constructor(
    val bitrate: Int? = null,
    val sampleRate: Int? = null,
    val echoCancellation: Boolean? = null,
    val noiseSuppression: Boolean? = null,
) {
    fun toMap(): Map<String, Any> =
        listOfNotNull(
            bitrate?.let { "bitrate" to it },
            sampleRate?.let { "sampleRate" to it },
            echoCancellation?.let { "echoCancellation" to it },
            noiseSuppression?.let { "noiseSuppression" to it },
        ).toMap()

    companion object {
        @JvmStatic
        fun fromMap(values: Map<String, Any>?): StreamAudioConfig? {
            values ?: return null
            return StreamAudioConfig(
                bitrate = numberValue(values, "bitrate"),
                sampleRate = numberValue(values, "sampleRate"),
                echoCancellation = values["echoCancellation"] as? Boolean,
                noiseSuppression = values["noiseSuppression"] as? Boolean,
            )
        }
    }
}

/** Effective video settings reported by the glasses after defaults and clamps. */
data class StreamResolvedVideoConfig @JvmOverloads constructor(
    /** Encoded output width sent to the stream endpoint. */
    val width: Int,
    /** Encoded output height sent to the stream endpoint. */
    val height: Int,
    /** Native camera buffer width selected before crop/downscale. */
    val captureWidth: Int? = null,
    /** Native camera buffer height selected before crop/downscale. */
    val captureHeight: Int? = null,
    /** Encoded video bitrate in bits per second. */
    val bitrate: Int,
    /** Resolved capture/encode frame rate. */
    val fps: Double,
) {
    fun toMap(): Map<String, Any> =
        buildMap {
            put("width", width)
            put("height", height)
            captureWidth?.let { put("captureWidth", it) }
            captureHeight?.let { put("captureHeight", it) }
            put("bitrate", bitrate)
            put("fps", fps)
        }

    companion object {
        @JvmStatic
        fun fromMap(values: Map<String, Any>?): StreamResolvedVideoConfig? {
            values ?: return null
            return StreamResolvedVideoConfig(
                width = numberValue(values, "width") ?: return null,
                height = numberValue(values, "height") ?: return null,
                captureWidth = numberValue(values, "captureWidth"),
                captureHeight = numberValue(values, "captureHeight"),
                bitrate = numberValue(values, "bitrate") ?: return null,
                fps = doubleValue(values, "fps") ?: return null,
            )
        }
    }
}

data class StreamResolvedAudioConfig @JvmOverloads constructor(
    val bitrate: Int? = null,
    val sampleRate: Int? = null,
    val echoCancellation: Boolean? = null,
    val noiseSuppression: Boolean? = null,
) {
    fun toMap(): Map<String, Any> =
        buildMap {
            bitrate?.let { put("bitrate", it) }
            sampleRate?.let { put("sampleRate", it) }
            echoCancellation?.let { put("echoCancellation", it) }
            noiseSuppression?.let { put("noiseSuppression", it) }
        }

    companion object {
        @JvmStatic
        fun fromMap(values: Map<String, Any>?): StreamResolvedAudioConfig? {
            values ?: return null
            return StreamResolvedAudioConfig(
                bitrate = numberValue(values, "bitrate"),
                sampleRate = numberValue(values, "sampleRate"),
                echoCancellation = boolValue(values, "echoCancellation"),
                noiseSuppression = boolValue(values, "noiseSuppression"),
            )
        }
    }
}

enum class StreamTransport(val value: String) {
    RTMP("rtmp"),
    SRT("srt"),
    WHIP("whip");

    companion object {
        @JvmStatic
        fun fromValue(value: String?): StreamTransport? =
            when (value?.lowercase()) {
                "rtmp" -> RTMP
                "srt" -> SRT
                "whip" -> WHIP
                else -> null
            }
    }
}

data class StreamResolvedConfig @JvmOverloads constructor(
    val transport: StreamTransport? = null,
    val video: StreamResolvedVideoConfig? = null,
    val audio: StreamResolvedAudioConfig? = null,
) {
    fun toMap(): Map<String, Any> =
        buildMap {
            transport?.let { put("transport", it.value) }
            video?.let { put("video", it.toMap()) }
            audio?.let { put("audio", it.toMap()) }
        }

    companion object {
        @JvmStatic
        fun fromMap(values: Map<String, Any>?): StreamResolvedConfig? {
            values ?: return null
            return StreamResolvedConfig(
                transport = StreamTransport.fromValue(stringValue(values, "transport")),
                video = StreamResolvedVideoConfig.fromMap(stringMapValue(values["video"])),
                audio = StreamResolvedAudioConfig.fromMap(stringMapValue(values["audio"])),
            )
        }
    }
}

/** Live encoder and device telemetry reported by the glasses while streaming. */
data class StreamLiveStats @JvmOverloads constructor(
    val bitrate: Long? = null,
    val fps: Double? = null,
    val droppedFrames: Long? = null,
    val duration: Long? = null,
    val temperatureC: Double? = null,
) {
    fun toMap(): Map<String, Any> =
        buildMap {
            bitrate?.let { put("bitrate", it) }
            fps?.let { put("fps", it) }
            droppedFrames?.let { put("droppedFrames", it) }
            duration?.let { put("duration", it) }
            temperatureC?.let { put("temperatureC", it) }
        }

    companion object {
        @JvmStatic
        fun fromMap(values: Map<String, Any>?): StreamLiveStats? {
            values ?: return null
            return StreamLiveStats(
                bitrate = longValue(values, "bitrate"),
                fps = doubleValue(values, "fps"),
                droppedFrames = longValue(values, "droppedFrames"),
                duration = longValue(values, "duration"),
                temperatureC = doubleValue(values, "temperatureC"),
            )
        }
    }
}

data class StreamRequest @JvmOverloads constructor(
    val streamUrl: String,
    val streamId: String = "",
    val sound: Boolean = true,
    val video: StreamVideoConfig? = null,
    val audio: StreamAudioConfig? = null,
    val authToken: String? = null,
) {
    fun toMap(): Map<String, Any> =
        buildMap {
            put("type", "start_stream")
            put("streamUrl", streamUrl)
            put("streamId", streamId)
            put("sound", sound)
            video?.toMap()?.takeIf { it.isNotEmpty() }?.let { put("video", it) }
            audio?.toMap()?.takeIf { it.isNotEmpty() }?.let { put("audio", it) }
            authToken?.takeIf { it.isNotEmpty() }?.let { put("authToken", it) }
        }

    companion object {
        @JvmStatic
        fun fromMap(values: Map<String, Any>): StreamRequest =
            StreamRequest(
                streamUrl =
                    (values["streamUrl"] ?: values["rtmpUrl"] ?: values["srtUrl"] ?: values["whipUrl"]) as? String
                        ?: "",
                streamId = values["streamId"] as? String ?: "",
                sound = values["sound"] as? Boolean ?: true,
                video = StreamVideoConfig.fromMap(stringMapValue(values["video"])),
                audio = StreamAudioConfig.fromMap(stringMapValue(values["audio"])),
                authToken = values["authToken"] as? String ?: values["auth_token"] as? String,
            )
    }
}

internal data class StreamKeepAliveRequest @JvmOverloads constructor(
    val streamId: String,
    val ackId: String,
) {
    fun toMap(): Map<String, Any> =
        mapOf(
            "type" to "keep_stream_alive",
            "streamId" to streamId,
            "ackId" to ackId,
        )

    companion object {
        @JvmStatic
        fun fromMap(values: Map<String, Any>): StreamKeepAliveRequest =
            StreamKeepAliveRequest(
                streamId = values["streamId"] as? String ?: "",
                ackId = values["ackId"] as? String ?: "",
            )
    }
}

enum class StreamState(val value: String) {
    INITIALIZING("initializing"),
    STREAMING("streaming"),
    STOPPING("stopping"),
    STOPPED("stopped"),
    RECONNECTING("reconnecting"),
    RECONNECTED("reconnected"),
    RECONNECT_FAILED("reconnect_failed"),
    ERROR("error");

    companion object {
        @JvmStatic
        fun fromValue(value: String?): StreamState? =
            when (value?.lowercase()) {
                "initializing", "starting", "connecting" -> INITIALIZING
                "streaming", "streaming_started", "active" -> STREAMING
                "stopping" -> STOPPING
                "stopped", "not_streaming", "disconnected", "timeout" -> STOPPED
                "reconnecting" -> RECONNECTING
                "reconnected" -> RECONNECTED
                "reconnect_failed" -> RECONNECT_FAILED
                "error", "error_not_streaming" -> ERROR
                else -> null
            }
    }
}

enum class StreamStatusKind(val value: String) {
    LIFECYCLE("lifecycle"),
    RECONNECT("reconnect"),
    ERROR("error"),
    SNAPSHOT("snapshot"),
}

sealed interface StreamStatus {
    val kind: StreamStatusKind
    val state: StreamState
    val streamId: String?
    val timestamp: Long?
    val resolvedConfig: StreamResolvedConfig?

    fun toMap(): Map<String, Any> {
        val values = mutableMapOf<String, Any>(
            "kind" to kind.value,
            "status" to state.value,
        )
        streamId?.takeIf { it.isNotBlank() }?.let { values["streamId"] = it }
        timestamp?.let { values["timestamp"] = it }
        resolvedConfig?.let { values["resolvedConfig"] = it.toMap() }

        when (this) {
            is Lifecycle -> Unit
            is Reconnecting -> {
                values["attempt"] = attempt
                values["maxAttempts"] = maxAttempts
                values["reason"] = reason
            }
            is Reconnected -> values["attempt"] = attempt
            is ReconnectFailed -> values["maxAttempts"] = maxAttempts
            is Error -> values["errorDetails"] = errorDetails
            is Snapshot -> {
                values["streaming"] = streaming
                values["reconnecting"] = reconnecting
                attempt?.let { values["attempt"] = it }
            }
        }

        return values
    }

    fun toEventMap(): Map<String, Any> = toMap() + mapOf("type" to "stream_status")

    data class Lifecycle(
        override val state: StreamState,
        override val streamId: String?,
        override val timestamp: Long?,
        override val resolvedConfig: StreamResolvedConfig?,
    ) : StreamStatus {
        override val kind: StreamStatusKind = StreamStatusKind.LIFECYCLE
    }

    data class Reconnecting(
        override val streamId: String?,
        val attempt: Int,
        val maxAttempts: Int,
        val reason: String,
        override val timestamp: Long?,
        override val resolvedConfig: StreamResolvedConfig?,
    ) : StreamStatus {
        override val kind: StreamStatusKind = StreamStatusKind.RECONNECT
        override val state: StreamState = StreamState.RECONNECTING
    }

    data class Reconnected(
        override val streamId: String?,
        val attempt: Int,
        override val timestamp: Long?,
        override val resolvedConfig: StreamResolvedConfig?,
    ) : StreamStatus {
        override val kind: StreamStatusKind = StreamStatusKind.RECONNECT
        override val state: StreamState = StreamState.RECONNECTED
    }

    data class ReconnectFailed(
        override val streamId: String?,
        val maxAttempts: Int,
        override val timestamp: Long?,
        override val resolvedConfig: StreamResolvedConfig?,
    ) : StreamStatus {
        override val kind: StreamStatusKind = StreamStatusKind.RECONNECT
        override val state: StreamState = StreamState.RECONNECT_FAILED
    }

    data class Error(
        override val streamId: String?,
        val errorDetails: String,
        override val timestamp: Long?,
        override val resolvedConfig: StreamResolvedConfig?,
    ) : StreamStatus {
        override val kind: StreamStatusKind = StreamStatusKind.ERROR
        override val state: StreamState = StreamState.ERROR
    }

    data class Snapshot(
        override val state: StreamState,
        val streaming: Boolean,
        val reconnecting: Boolean,
        override val streamId: String?,
        val attempt: Int?,
        override val timestamp: Long?,
        override val resolvedConfig: StreamResolvedConfig?,
    ) : StreamStatus {
        override val kind: StreamStatusKind = StreamStatusKind.SNAPSHOT
    }

    companion object {
        @JvmStatic
        fun fromMap(values: Map<String, Any>): StreamStatus {
            val rawState = stringValue(values, "status")
            val streaming = boolValue(values, "streaming")
            val reconnecting = boolValue(values, "reconnecting") ?: false
            val streamId = stringValue(values, "streamId")
            val timestamp = longValue(values, "timestamp")
            val resolvedConfig = StreamResolvedConfig.fromMap(stringMapValue(values["resolvedConfig"]))
            val attempt = numberValue(values, "attempt")
            val maxAttempts = numberValue(values, "maxAttempts") ?: 0
            val parsedState = StreamState.fromValue(rawState)

            if (streaming != null || hasAnyKey(values, "reconnecting")) {
                return Snapshot(
                    state = when {
                        reconnecting -> StreamState.RECONNECTING
                        streaming == true -> StreamState.STREAMING
                        parsedState != null -> parsedState
                        else -> StreamState.STOPPED
                    },
                    streaming = streaming == true,
                    reconnecting = reconnecting,
                    streamId = streamId,
                    attempt = attempt,
                    timestamp = timestamp,
                    resolvedConfig = resolvedConfig,
                )
            }

            val state = parsedState
                ?: return Error(
                    streamId = streamId,
                    errorDetails = rawState?.let { "Unknown stream status: $it" } ?: "Missing stream status",
                    timestamp = timestamp,
                    resolvedConfig = resolvedConfig,
                )

            return when (state) {
                StreamState.RECONNECTING -> Reconnecting(
                    streamId = streamId,
                    attempt = attempt ?: 0,
                    maxAttempts = maxAttempts,
                    reason = stringValue(values, "reason") ?: "",
                    timestamp = timestamp,
                    resolvedConfig = resolvedConfig,
                )
                StreamState.RECONNECTED -> Reconnected(
                    streamId = streamId,
                    attempt = attempt ?: 0,
                    timestamp = timestamp,
                    resolvedConfig = resolvedConfig,
                )
                StreamState.RECONNECT_FAILED -> ReconnectFailed(
                    streamId = streamId,
                    maxAttempts = maxAttempts,
                    timestamp = timestamp,
                    resolvedConfig = resolvedConfig,
                )
                StreamState.ERROR -> Error(
                    streamId = streamId,
                    errorDetails = stringValue(values, "errorDetails")
                        ?: if (rawState == "error_not_streaming") "not_streaming" else "Unknown stream error",
                    timestamp = timestamp,
                    resolvedConfig = resolvedConfig,
                )
                else -> Lifecycle(
                    state = state,
                    streamId = streamId,
                    timestamp = timestamp,
                    resolvedConfig = resolvedConfig,
                )
            }
        }
    }
}

data class StreamStatusEvent(
    val status: StreamStatus,
    val stats: StreamLiveStats? = null,
) {
    // True when the glasses will retry the failed publisher themselves
    // (emitting side lands in PR #3488); absent on older firmware and on
    // events not parsed from a glasses status map. Carried here instead of
    // on StreamStatus.Error so the public Error shape stays unchanged.
    var willRetry: Boolean? = null
        private set

    constructor(values: Map<String, Any>) : this(
        status = StreamStatus.fromMap(values),
        stats = StreamLiveStats.fromMap(stringMapValue(values["stats"])),
    ) {
        willRetry = boolValue(values, "willRetry")
    }

    val state: StreamState get() = status.state
    val streamId: String? get() = status.streamId
    val resolvedConfig: StreamResolvedConfig? get() = status.resolvedConfig
    val values: Map<String, Any>
        get() = buildMap {
            putAll(status.toEventMap())
            stats?.let { put("stats", it.toMap()) }
            willRetry?.let { put("willRetry", it) }
        }
}

data class KeepAliveAckEvent(
    val streamId: String,
    val ackId: String,
    val timestamp: Long?,
) {
    constructor(values: Map<String, Any>) : this(
        streamId = stringValue(values, "streamId").orEmpty(),
        ackId = stringValue(values, "ackId").orEmpty(),
        timestamp = longValue(values, "timestamp"),
    )

    val values: Map<String, Any>
        get() = buildMap {
            put("type", "keep_alive_ack")
            put("streamId", streamId)
            put("ackId", ackId)
            timestamp?.let { put("timestamp", it) }
        }
}
